src/python-common/ceph/deployment/service_spec.py

```diff
@@ -1095,3 +1113,106 @@ def get_port(self) -> int:
+                 preview_only: bool = False,
+                 port: Optional[int] = 9464,
+                ):
+        assert service_type == 'snmp-gateway'
+        assert snmp_version is not None
+        assert snmp_destination is not None
+        assert credentials is not None
+
+        super(SNMPGatewaySpec, self).__init__(
+            service_type,
+            placement=placement,
+            unmanaged=unmanaged,
+            preview_only=preview_only)
+
+        self.service_type = service_type
+        self.snmp_version = snmp_version
+        self.snmp_destination = snmp_destination
+        self.port = port
+        self.credentials = credentials
+        self.engine_id = engine_id
+        self.auth_protocol = auth_protocol
+        self.privacy_protocol = privacy_protocol
+
+    @property
+    def ports(self) -> List[int]:
+        return [self.port]
+
+    def get_port_start(self) -> List[int]:
+        return self.ports
+
+    def validate(self) -> None:
+        super(SNMPGatewaySpec, self).validate()
+
+        def _check_type(name: str, value: Optional[str], options: List[str]) -> None:
+            if not value:
+                return
+            if value not in options:
+                raise SpecValidationError(
+                    f'{name} unsupported. Must be one of {", ".join(options)}'
+                )
+
+        _check_type('snmp_version', self.snmp_version, list(set(opt.value for opt in SNMPVersion)))
+        _check_type('auth_protocol', self.auth_protocol, list(set(opt.value for opt in SNMPAuthType)))
+        _check_type('privacy_protocol', self.privacy_protocol, list(set(opt.value for opt in SNMPPrivacyType)))
+
+        creds_requirement = {
+            'V2c': ['snmp_community'],
+            'V3': ['snmp_v3_auth_username', 'snmp_v3_auth_password']
+        }
+        if self.privacy_protocol:
+            creds_requirement['V3'].append('snmp_v3_priv_password')
+
+        missing = [parm for parm in creds_requirement[self.snmp_version] if parm not in self.credentials]
+        # check that credentials are correct for the version
+        if missing:
+            raise SpecValidationError(
+                f"SNMP {self.snmp_version} credentials are incomplete. Missing {', '.join(missing)}"
+            )
+
+        if self.engine_id:
+            if 10 >= len(self.engine_id) >= 64 or \
+               not is_hex(self.engine_id):
+                raise SpecValidationError(
+                    "engine_id must be a hex string, between 10-64 chars in length"
+                )
+
+        else:
+            if self.snmp_version == 'V3':
+                raise SpecValidationError(
+                    "must provide an engine_id for SNMP v3 notifications"
+                )
+
+        if not self.snmp_destination:
+            raise SpecValidationError(
+                "SNMP destination (snmp_destination) must be provided"
+            )
+        else:
+            dest = self.snmp_destination.split(':')
```

Comment:
broken for ipv6

Reply:
Agree. Is there a preferred pattern or existing function for dealing with ipv6 in the orch/cephadm code? I've been focussed on ipv4 :( For ipv6, should the expected format be  [ipv6]:port for example as per http://www.ietf.org/rfc/rfc3986.txt?

Reply:
yes: https://github.com/ceph/ceph/commit/d65ccf3a0f83fc4a79edfe46c9468cf77f655f83